{% extends "base.html" %} {% block page_title %}
<h1 class="title">Meta-domains</h1>
<h2 class="subtitle">Make use of domain-homology to interpret
  genetic variants</h2>
{% endblock %} {% block content %}

<!-- Scripts needed for functionality on this page -->
<script src="{{ url_for('web.metadom_js') }}"></script>

<!-- Custom stylesheet for this page -->

<!-- Staring definition of site content -->
<section class="section">
  <div class="container">
    <div class="label"><label class="title">Meta-domain information in the context of {{data['meta_domain_info']['gene_name']}}</label></div> 
    <div class="label">
      <label class="label">
        Gene: {{data['meta_domain_info']['gene_name']}} (
        <a href="http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?t={{data['meta_domain_info']['transcript']}}" target="_blank">
          {{data['meta_domain_info']['transcript']}}
        </a>), strand: {{data['meta_domain_info']['strand']}},  cDNA length: {{data['meta_domain_info']['cDNA_length']}}
      </label>
    </div>
    <div class="label">
      <label class="label">
        Protein {{data['meta_domain_info']['protein_name']}} (
        <a href="http://www.uniprot.org/uniprot/{{data['meta_domain_info']['protein_ac']}}" target="_blank">
          {{data['meta_domain_info']['protein_ac']}}
        </a>) protein length: {{data['meta_domain_info']['protein_length']}}
      </label>
    </div>
    <div class="label">
      <label class="label">
        Domain 
        <a href="http://pfam.xfam.org/family/{{data['meta_domain_info']['domain_id']}}" target="_blank">
          {{data['meta_domain_info']['domain_id']}}
        </a>, with {{data['meta_domain_info']['domain_occurrences']|length}} occurrences in this protein at {% for domain_occurrence in data['meta_domain_info']['domain_occurrences'] %} ({{domain_occurrence['start']}}-{{domain_occurrence['stop']}})  {% endfor %}
      </label>
    </div>
    
    <div class="label"><label class="title">Meta-domain data in the context of {{data['meta_domain_info']['gene_name']}}</label></div> 
    <table class="table is-hoverable is-fullwidth">
      <thead>
        <tr>
          <th><abbr title="Position in protein">Protein position</abbr></th>
          <th><abbr title="Chromosome positions for the codon">Chromosome positions</abbr></th>
          <th><abbr title="Position based on the Pfam protein domain">Consensus position</abbr></th>
          <th><abbr title="Percentage of positions that can be aligned throughout the human genome">Alignment %</abbr></th>
          <th><abbr title="Relative entropy as measured via Pfam full alignments">Pfam entropy</abbr></th>
          <th><abbr title="Relative entropy as measured solely via homology relations in the human genome, incorporating ExAC variants">ExAC entropy</abbr></th>
          <th><abbr title="Number of missense variants found in ExAC at this domain position">Meta ExAC missense </abbr></th>
          <th><abbr title="Number of missense variants found in HGMD at this domain position">Meta HGMD missense </abbr></th>
          <th><abbr title="Identifiers for the variants found in ExAC at this domain position">ExAC identifiers</abbr></th>
          <th><abbr title="Identifiers for the variants found in HGMD at this domain position">HGMD identfiers </abbr></th>
        </tr>
      </thead>
      <tfoot>
        <tr>
          <th><abbr title="Position in protein">Protein position</abbr></th>
          <th><abbr title="Chromosome positions for the codon">Chromosome positions</abbr></th>
          <th><abbr title="Position based on the Pfam protein domain">Consensus position</abbr></th>
          <th><abbr title="Percentage of positions that can be aligned throughout the human genome">Alignment %</abbr></th>
          <th><abbr title="Relative entropy as measured via Pfam full alignments">Pfam entropy</abbr></th>
          <th><abbr title="Relative entropy as measured solely via homology relations in the human genome, incorporating ExAC variants">ExAC entropy</abbr></th>
          <th><abbr title="Number of missense variants found in ExAC at this domain position">Meta ExAC missense </abbr></th>
          <th><abbr title="Number of missense variants found in HGMD at this domain position">Meta HGMD missense </abbr></th>
          <th><abbr title="Identifiers for the variants found in ExAC at this domain position">ExAC identifiers</abbr></th>
          <th><abbr title="Identifiers for the variants found in HGMD at this domain position">HGMD identfiers </abbr></th>
        </tr>
      </tfoot>
      <tbody id="domainTable">
      {% for meta_domain_entry in data['meta_domain_data'] %}
        <tr id="row{{meta_domain_entry['protein_pos']}}" onclick='$("[id^=row]").removeClass("is-selected");$("#row{{meta_domain_entry['protein_pos']}}").addClass("is-selected")'>
        <th>{{meta_domain_entry['protein_pos']}}</th>        
        <td>{{meta_domain_entry['consensus_pos']}}</td>
        <td>{{meta_domain_entry['chr_region']}}</td>
        <td>{{meta_domain_entry['codon']}}</td>
        <td>{{meta_domain_entry['amino_acid']}}</td>
        <td>{{meta_domain_entry['other_chr_regions']}}</td>
        <td>{{meta_domain_entry['other_codons']}}</td>
        <td>{{meta_domain_entry['other_amino_acids']}}</td>
        <td>{{meta_domain_entry['other_normal_variation']}}</td>
        <td>{{meta_domain_entry['other_pathogenic_variation']}}</td>
        </tr> 
      {% endfor %}
      </tbody>
    </table>
    <div class="field">
      <div class="field has-addons">
        <div class="control">
          <button class="button is-info" onclick="clearTable()">
            Clear Table</button>
        </div>
      </div>
    </div>
  </div>
</section>

{% endblock %} {% block js %} {% endblock %}
