{% extends "base.html" %} {% block content %}

<!-- Staring definition of site content -->
<section class="section">
  <div class="box">
    <section class="section">
      <div class='content'>
        <h1 class="title" style="color: #00AFDC;">Frequently Asked Questions</h1>
        
        <p>
          If your question is not here, please do not hesitate to <a
            href="{{ url_for('web.contact') }}">contact us</a>.
        </p>
        
        <h3 class="title" style="color: #00AFDC;">General questions</h3>
        <div class="box">
	        <div class="label">What is MetaDome?</div>
	        <div class="box">
	        	MetaDome is a fast and easy-to-use web server that offers an easily 
	        	accessible and unique perspective that enables researchers and 
	        	clinicians to interpret variants of unknown significance via 
	        	'meta-domains'; a concept that we have previously validated 
	        	(<a href="https://doi.org/10.1002/humu.23313" target="_blank">
          		L. Wiel et al. Human Mutation, 2017 </a>) which makes use of protein 
          		domain homology within the human genome to transfer genetic variant 
          		information across homologous domains. MetaDome can visualize 
          		meta-domain information and gene-wide profiles of genetic tolerance.
	        </div>
	        
	        
	        <div class="label">How can I use MetaDome?</div>
	        <div class="box">
				Analyses through MetaDome can be performed by querying a gene name 
				and desired transcript. The result is a visualization that a user 
				may interact with to highlight positions of their interest and 
				obtain more detailed information of those positions. This includes 
				regional tolerance to missense variation and the presence or absence 
				of variation at corresponding positions within homologous domains.
				<br><br>
				To get more familiar with the using MetaDome and all of it's functionality 
				we strongly advise to <a href="{{ url_for('web.dashboard') }}?tour=true">
				start the tour</a>.
	        </div>
	        
	        <div class="label">Can I use my smartphone or tablet to access MetaDome?</div>
	        <div class="box">
	        	Although we recommend using a laptop/desktop with a compatible 
	        	browser for the best user experience, we designed the user interface 
	        	to be useable for mobile devices as well. If you encounter any problems or 
	        	see any possible improvements please  
	        	<a href="{{ url_for('web.contact') }}">contact us</a>. 
	        </div>
	         
	        <div class="label">What desktop browsers are compatible for MetaDome?</div>
	        <div class="box">
		        <p>
		          The MetaDome web server is developed and extensively tested using 
		          <a href="https://www.google.com/chrome/">Google Chrome</a> and 
		          <a href="https://www.mozilla.org/en-US/firefox/new/">Firefox</a>. We 
		          recommend using one of these browsers to access all functionality. 
		          Here is a list of compatible and tested browsers:
		        </p>
		        <table class="table">
				  <thead>
				    <tr>
				      <th><abbr title="OS">OS</abbr></th>
				      <th><abbr title="Version">Version</abbr></th>
				      <th><abbr title="Chrome">Chrome</abbr></th>
				      <th><abbr title="Firefox">Firefox</abbr></th>
				      <th><abbr title="Microsoft Edge">Microsoft Edge</abbr></th>
				      <th><abbr title="Microsoft IE">Microsoft IE</abbr></th>
				      <th><abbr title="Safari">Safari</abbr></th>
				    </tr>
				  </thead>
				  <tbody>
				    <tr>
				      <th>Linux</th>
				      <td>Ubuntu 16.04 and 18.04</td>
				      <td>70.0</td>
				      <td>63.0</td>
				      <td>n/a</td>
				      <td>n/a</td>
				      <td>n/a</td>
				    </tr>
				    <tr>
				      <th>MacOS</th>
				      <td>Mojave</td>
				      <td>70.0</td>
				      <td>63.0</td>
				      <td>n/a</td>
				      <td>n/a</td>
				      <td>Not supported</td>
				    </tr>
				    <tr>
				      <th>Windows</th>
				      <td>7 and 10</td>
				      <td>70.0</td>
				      <td>63.0</td>
				      <td>Not tested</td>
				      <td>Not supported</td>
				      <td>n/a</td>
				    </tr>
				  </tbody>
				</table>
	        </div>
	        
	        <div class="label">I have ran into a problem with MetaDome, what can I do?</div>
			<div class="box">
				Although we have extensively tested MetaDome and are constantly monitoring 
				the server for any downtime. If you do encounter problems or behaviour you
				did not expect. We would highly appreciate it if you could 
				<a href="{{ url_for('web.contact') }}">contact us</a>.
			</div>
			
	        <div class="label">On what data are the results in MetaDome based?</div>
			<div class="box">
				MetaDome combines data from various public resources. The underlying data 
				of MetaDome currently contains information from 56,319 human transcripts 
				(<a href="https://www.gencodegenes.org/" target="_blank">GENCODE</a>), 
				71,419 <a href="https://pfam.xfam.org/" target="_blank">Pfam</a> 
				protein domain instances, 12,164,292 population-based genetic variants from 
				<a href="http://gnomad.broadinstitute.org/">gnomAD</a>, and 34,076 pathogenic mutations from 
				<a href="https://www.ncbi.nlm.nih.gov/clinvar/" target="_blank">ClinVar</a>. 
			</div>
			
			<div class="label">How often is the underlying data updated?</div>
			<div class="box">
				Currently MetaDome requires a manual reannotation of the underlying data.
				In a future version of MetaDome is expected to schedule automatic downloading 
				of new data releases and automatic reannotation.
			</div>
			
			
	        <div class="label">Is MetaDome still under construction?</div>
			<div class="box">
				MetaDome is final for its first release version. But we are still actively 
				attempting to improve MetaDome based on feedback. If you have any improvement 
				ideas, please <a href="{{ url_for('web.contact') }}">contact us</a>.
				Current improvements that we are working on may be found 
				<a href="https://github.com/cmbi/metadome/issues">here</a>. 
			</div>
			
			<div class="label">Why can I not select my gene / transcript for analysis?</div>
			<div class="box">
				For sake of high data quality we have limited the possible transcripts and genes 
				suitable for a MetaDome Analysis to the <a href="https://www.gencodegenes.org/" target="_blank">GENCODE</a> 
				Basic transcripts of which the translation is identically matched with a 
				<a href="http://www.uniprot.org/" target="_blank">SwissProt</a> protein 
				sequence. All GENCODE Basic transcripts are validated to be expressed as mRNA 
				and all SwissProt proteins are curated by experts.
			</div>
			
			<div class="label">I cannot find the domain that should be in my gene of interest?</div>
			<div class="box">
				All annotated protein domains in MetaDome result from 
				<a href="https://pfam.xfam.org/" target="_blank">Pfam</a>. 
				Pfam holds a strict notion on what a protein domain is and differs to 
				other protein domain annotation services. We have specifically chose Pfam, 
				as the underlying method is suitable to indicate homologously identical 
				positions across protein domains that can directly be used to transfer variant 
				information. 
			</div>

        </div>
        
        <h3 class="title" style="color: #00AFDC;">Interpretation of a MetaDome analysis</h3>      
        <div class="box">
			<div class="label">Why are there black boxes in the meta-domain landscape?</div>
			<div class="box">
				The black boxes represent positions that cannot be aligned to the Pfam 
				domain consensus. For these positions we 
				cannot transfer variant information across homologous domains.
			</div>
			
			<div class="label">How is the tolerance landscape computed?</div>
			<div class="box">
				The tolerance is computed as a from missense over synonymous 
				variant count ratio, which is calculated in a sliding window manner to 
				provide a per-position indication of regional tolerance to missense variation.
				The variants are based on gnomAD and corrected for codon composition. 
				For more information we suggest reading 
				<a href="{{ url_for('web.method') }}#tolerance_computation">"Computing genetic 
				tolerance and generating a tolerance landscape"</a>.
			</div>

			<div class="label">What do the tolerance landscape' colours indicate?</div>
			<div class="box">
				The colours are based on missense over synonymous ratio computations 
				over the entire protein for all proteins. Here the Yellow (or neutral) 
				is based on the average missense over synonymous ratio score that you 
				would obtain this way and the red (intolerant) and blue (tolerant) are 
				based on the top/bottom scores.
			</div>

			<div class="label">For meta-domain information in the positional information, what does the alignment coverage mean?</div>
			<div class="box">
				This corresponds to how many homologous positions are aligned to that 
				specific position. In general we would advise to maintain a minimum of 
				80% alignment coverage for high quality results.
			</div>
		</div>
      </div>
    </section>
  </div>
</section>

{% endblock %} {% block js %} {% endblock %}
